src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1147,9 +1147,32 @@ async ValueTask<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, b
                 if (conn.TryGetBoundConnector(out var connector))
                 {
                     connector.StartUserAction(this);
+
+                    CancellationTokenSource? cts = null;
+                    CancellationTokenRegistration? registration = null;
+                    var finalCt = CancellationToken.None;
+                    var cancellationTimeout = connector.Settings.CancellationTimeout;
+
                     try
                     {
-                        using var _ = cancellationToken.Register(cmd => ((NpgsqlCommand)cmd!).Cancel(), this);
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            cts = new CancellationTokenSource();
+                            registration = cancellationToken.Register(cmd =>
+                            {
+                                try
+                                {
+                                    ((NpgsqlCommand)cmd!).Cancel(true);
+                                    if (cancellationTimeout > 0)
```

Comment:
Just a thought (not for this PR)... It's true that 0 usually means "no timeout" (e.g. with CommandTimeout). However, CancellationTimeout is a bit special, in that a zero timeout could actually make sense - it would mean "don't attempt to send a PG cancellation and just break the connection right away.

I'm not sure if disabling PG cancellation could actually be useful to anyone - maybe some pseudo-PostgreSQL variant that doesn't support cancellation. It's also a bit confusing to give 0 a different meaning for CancellationTimeout than what it means for CommandTimeout. But something makes me want to support this, for completeness (and possibly use -1 to mean no timeout).

Any thoughts @vonzshik @YohDeadfall @Brar?

(no need to block this PRs on this question, we can change this later).

Reply:
An interesting idea. And I don't a reason why not - although I highly doubt anyone would really use it. Maybe only in a case, when you want the fastest shutdown possible.